docs/research/ha/RATIONALE.md

```diff
@@ -0,0 +1,446 @@
+model across all the pipelines being part of the same logical service.
+
+Under these conditions, a client accessing the S3 service would experience a failure
+only for *the active operation* that were happening on the failed s3gw pipeline.
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw) ==> PV((PV))
+    ingress ==> radosgw2(radosgw) ==> PV2((PV))
+
+    linkStyle 0 stroke: red
+    linkStyle 1 stroke: red
+    linkStyle 2 stroke: red
+```
+
+The subsequent access to the S3 service would result in a successful operation
+because the load balancer would immediately reroute to another active pipeline.
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw) ==> PV((PV))
+    ingress ==> radosgw2(radosgw) ==> PV2((PV))
+
+    linkStyle 0 stroke: green
+    linkStyle 1 stroke: red
+    linkStyle 2 stroke: red
+    linkStyle 3 stroke: green
+    linkStyle 4 stroke: green
+```
+
+This model is potentially hugely expensive for the s3gw project
+because there isn't an implementation for the data replication with the SFS backend.
+
+## Active-Passive pipelines
+
+This model assumes N pipelines to be "allocated" on the cluster at the same time,
+but only one, the *active* pipeline, owns the *exclusive ownership*
+over the shared `PV`.
+
+Should the active pipeline suffer a failure, the next elected active pipeline,
+chosen to replace the old one, should pay a *time window* necessary to "transfer"
+the ownership over the shared `PV`.
+
+This means that, unlike the *Active-Active* solution, it is not guaranteed that
+the subsequent access made by a client after a failed one is always successful.
+
+It can happen that a client issues an access to the S3 service while the system is
+translating the passive pipeline into the active state; in this case, the issued
+access is expected to fail.
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw) ==> PV((PV))
+    ingress ==> radosgw2(radosgw)
+
+    linkStyle 0 stroke: red
+    linkStyle 1 stroke: red
+    linkStyle 2 stroke: red
+```
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw)
+    ingress ==> radosgw2(radosgw) == ownership transfer ==> PV((PV))
+
+    linkStyle 0 stroke: red
+    linkStyle 1 stroke: white
+    linkStyle 2 stroke: red
+```
+
+```mermaid
+flowchart LR
+    client(client) ==> ingress
+    ingress(ingress) ==> radosgw(radosgw)
+    ingress ==> radosgw2(radosgw) ==> PV((PV))
+
+    linkStyle 0 stroke: green
+    linkStyle 1 stroke: white
+    linkStyle 2 stroke: green
+    linkStyle 3 stroke: green
+```
+
+## One Active pipeline
+
+This model is logically equivalent to the *Active-Passive* model with the difference
+that only one pipeline is allocated on the cluster at a time.
+
+In the *One pipeline* model, the only notable perceived difference between
```

Comment:
Again, I argue this isn't true. The fail-over latency is very unlikely to be measurably worse (there's nothing meaningful preloaded in the second model), while at the same time, the fault-free operation performance is going to be better, and the complexity is much lower.